src/asset_proofs/const_time_elgamal_encryption.rs

```diff
@@ -0,0 +1,207 @@
+//! In this implementation, we set `random_1` to the blinding factor used for the
+//! twisted Elgamal encryption. This way the twisted Elgamal and regular Elgamal
+//! ciphertexts can share the same `X`.
+
+use bulletproofs::PedersenGens;
+use curve25519_dalek::{
+    ristretto::{CompressedRistretto, RistrettoPoint},
+    scalar::Scalar,
+};
+use rand_core::{CryptoRng, RngCore};
+
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+use codec::{Decode, Encode, Error as CodecError, Input, Output};
+use sha3::{digest::FixedOutput, Digest, Sha3_256};
+use sp_std::prelude::*;
+
+use crate::asset_proofs::elgamal_encryption::{
+    CipherText, CommitmentWitness, ElgamalPublicKey, ElgamalSecretKey,
+};
+
+use crate::errors::Fallible;
+
+/// This data structure wraps a twisted Elgamal cipher text with the
+/// regular Elgamal cipher text.
+/// Since regular Elgamal decryption is constant time, its result is
+/// used as a hint to verify the twisted elgamal encryption.
+/// Note that we can not only rely on regular Elgamal encryption since
+/// 1. it is not homomorphic. 2. all asset proofs prove properties of
+/// a twisted Elgamal cipher text.
+#[derive(PartialEq, Copy, Clone, Default)]
+#[cfg_attr(feature = "std", derive(Debug))]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct CipherTextWithHint {
+    // The twisted Elgamal cipher text.
+    pub elgamal_cipher: CipherText,
+
+    pub y: RistrettoPoint,
+    pub z: [u8; 32],
+}
+
+impl Encode for CipherTextWithHint {
+    #[inline]
+    fn size_hint(&self) -> usize {
+        self.elgamal_cipher.size_hint() + 2 * 32
+    }
+
+    fn encode_to<W: Output>(&self, dest: &mut W) {
+        let y = self.y.compress();
+
+        self.elgamal_cipher.encode_to(dest);
+        y.as_bytes().encode_to(dest);
+        self.z.encode_to(dest);
+    }
+}
+
+impl Decode for CipherTextWithHint {
+    fn decode<I: Input>(input: &mut I) -> Result<Self, CodecError> {
+        let (elgamal_cipher, y, z) = <(CipherText, [u8; 32], [u8; 32])>::decode(input)?;
+        let y = CompressedRistretto(y)
+            .decompress()
+            .ok_or_else(|| CodecError::from("CipherTextWithHint Y point is invalid"))?;
+
+        Ok(CipherTextWithHint {
+            elgamal_cipher,
+            y,
+            z,
+        })
+    }
+}
+
+// ------------------------------------------------------------------------
+// Constant Time Elgamal Encryption
+// ------------------------------------------------------------------------
+
+// Generate a one-time-pad from Hash(key) and byte-wise xor it with the data.
+fn xor_with_one_time_pad(key: RistrettoPoint, data: &[u8; 32]) -> [u8; 32] {
+    let key_bytes: [u8; 32] = key.compress().to_bytes();
+    let hashed_key = Sha3_256::default().chain(key_bytes).fixed_result();
+
+    let mut result = [0u8; 32];
+    for index in 0..32 {
+        result[index] = hashed_key[index] ^ data[index];
+    }
+
+    result
```

Comment:
```suggestion
hashed_key
    .iter()
    .zip(data.iter())
    .map(|(&x1, &x2)| x1 ^ x2)
    .collect()
```

Reply:
Yeah, using zip and map is more Rust-like, the problem is `collect()` won't collect into a fixed size array, it can only collect into a `Vec<u8>` then I'll need to copy that into a result array if I go with this.